### ISSUE TITLE
Problem with providing jars as part of spark loading thru sparklyr spark_connect and loading properly

### ISSUE THREAD

**Issue by ying1** [sparklyr/sparklyr#2776]
This is an issue that was observed as a change from sparklyr 1.0.2 to sparklyr 1.4.0 

Previously, with sparklyr 1.0.2 - a connection w/ livy jars can be done this way, and a class that is contained in the jar file can be loaded without issues.  NOTE: the code is not using the `version` tag when using `spark_connect`
```
 spark_master <- "https:/mysystem:8998/livy2/v1"
 method <- "livy"
 custom_headers <- c(custom_headers, "Authorization"= btok)

 conf <- sparklyr::livy_config(custom_headers = custom_headers, proxy_user=proxyuser)
 conf$livy.jars <- c(
   "/user/user1/lib/myjar1.jar",
   "/user/user1/lib/myjar2.jar")

  conf$spark.sql.catalogImplementation <- NULL
 conf$livy.session.start.timeout=3000
 options(sparklyr.verbose = TRUE)
  sc <- spark_connect(master = spark_master, config=conf, method=method)

  sparklyr::invoke_static(sc, "com.ibm.connect.MyClass", "myMethod")
```

When we pick up sparklyr 1.4.0 - the above does not work anymore.  

The spark_connect method now requires a version, in which some CDH version fails to work (such as: `2.4.0-cdh6.1.1`).  The main problem is that  jars that are needed does not seem to be passed or cannot be loaded anymore.  NOTE: same jars has no issues when connecting via `sparklyr 1.0.2`

I had tested with a few various combinations, but still am not able to have this run correctly.
1. Replacing `livy.jars` w/ `jars` does not pass the jars to livy.
2. Set in addition, a scala version does not work, also does not pass jars to livy
There are a few more things that I had tested, but I do not see the jar files passed to livy. 

The only sort of "success" was that - I can see the list of jars passed to livy during the session creation via this method, but then that causes the exception below.
```
conf$livy.jars <- c( ... )
conf$sparklyr.livy.sources <- TRUE
```

Err:
```
Error in h(simpleError(msg, call)) :
 Failed to initialize livy connection: Failed to execute Livy statement with error: <console>:24: error: not found: value LivyUtils
Traceback:        var sparklyrRetVar_1 = LivyUtils.invokeFromBase64("AAAAHm9yZy5hcGFjaGUuc3BhcmsuU3BhcmtDb250ZXh0AAAAAAEAAAAMZ2V0T3JDcmVhdGUAAAAAAA==")
```

This is a separate question.   When looking at livy logs, I see an extra jar file appended in the list `https://github.com/sparklyr/sparklyr/blob/feature/sparklyr-1.3.0/inst/java/sparklyr-2.3-2.11.jar?raw=true` 
Can this be avoid?  On the system that we are working with, it is not guarentee that this system can access external github.

**Comment by yitao-li** [comment]
@ying1 I don't think `sparklyr` supports downloading required the jar file from a custom URL yet. It should be something that is reasonably easy to implement though.

Re: `sparklyr.livy.sources` -- I think it's something that was supported in older versions of Spark but not supported in Spark 2.4 (as evident from the error message you ran into).

@javierluraschi should we remove all code paths related to `sparklyr.livy.sources` completely or just the ones already deleted in https://github.com/sparklyr/sparklyr/pull/2234? It's unclear what's the current status of it, and whether it will work at all after numerous the changes made to scala source files.

**Comment by ying1** [comment]
@yitao-li 
What is the correct way to load jars as part of livy start up?  
It is not clear to me why these jars are needed `sparklyr-2.3-2.11.jar`

**Comment by yitao-li** [comment]
@ying1 It looks like there is a bug that stopped additional custom jar files from being loaded. I'm working on the fix.

Re: `sparklyr-2.3-2.11.jar` -- it's needed because `sparklyr` needs to launch a Spark application that essentially works as a server for `sparklyr` on Spark side accepting requests from functions such as `spark_connect()`, `invoke()/invoke_static()`, and `spark_disconnect()`.

**Comment by ying1** [comment]
@yitao-li 

> Re: sparklyr.livy.sources
This just happens to be a flag that I noticed in the code that I can set - without this line of the code overwriting livy.jars configuration (let me see if I remember where I saw the code) 

I did also try to use `conf$sparklyr.livy.jar` - but that ended up w/ some syntax error.
```
conf$sparklyr.livy.jar <- c(...) 

Err:
> sc <- spark_connect(master = spark_master, config=conf, method=method, version="2.3.2.3.1.4.0-315")
Error in livy_validate_http_response("Failed to create livy session",  : 
  Failed to create livy session (Client error: (400) Bad Request): {"msg":"java.net.URISyntaxException: Illegal character in path at index 2: c(\"/user/dsxhi/lib/spark/connectors/dataconn-engine-spark-status-3.0.1171-SNAPSHOT.jar\", \"/user/dsxhi/lib/spark/connectors/dataconn-engine-spark-udf-3.0.1167.jar\", \"/user/dsxhi/lib/spark/connectors/gson-2.8.5.jar\",  ... )"}
```

**Comment by ying1** [comment]
@yitao-li 

> Re: sparklyr-2.3-2.11.jar 

Is this jar specific to Spark version? 

Is this something that can be downloaded on a remote system, and specified to be used?

**Comment by yitao-li** [comment]
@ying1 I don't think local file paths are supported by Livy (as seen in the error message, Livy is trying to parse that file path as a URI).

Re: `sparklyr-2.3-2.11.jar` Yes we follow a naming convention for those files. `sparklyr-2.3-2.11.jar` is specific to Spark version 2.3.x built with Scala compiler version 2.11

**Comment by ying1** [comment]
> I don't think local file paths are supported by Livy (as seen in the error message, Livy is trying to parse the string value as a URI, so something like "https://my-server/file" would work). 

So you mean for `conf$sparklyr.livy.jar` - this expects an URI ?  or ? Is this different than `conf$livy.jars` ?  The documentation was not clear what these fields mean.

Also, when using `sparklyr-1.0.2` using Livy - referencing jars that are in HDFS (via `conf$livy.jars`), the jar list does not require to be URI 

```
20/10/20 11:50:29 INFO InteractiveSession: Stopped InteractiveSession 31.
20/10/20 12:11:17 INFO InteractiveSession$: Creating Interactive session 32: [owner: dsxhi, request: [kind: spark, proxyUser: Some(user1), jars: /user/user1/lib/myjar.jar,/user/user1/lib/myjar2.jar, ... , conf: spark.yarn.queue -> default,spark.driver.memory -> 1024m,spark.executor.instances -> 1,spark.driver.cores -> 1,spark.executor.memory -> 1024m,spark.executor.cores -> 1, heartbeatTimeoutInSecond: 0]]
```

>  Re: sparklyr-2.3-2.11.jar

Thank you.  This means, we'll need to match the jar w/ whatever spark & scala version we have?

**Comment by yitao-li** [comment]
@ying1 Yes that's correct.

**Comment by ying1** [comment]
@yitao-li 

Thank you. To summarize:

The main issue re: livy.jars are not loaded.  This one you mention is a bug. 

In addition to the jars for livy,  the `sparklyr-<spark_ver>-<scala_ver>.jar `will need to be included as part of this list.  This sparklyr.jar is required depending on the version of spark/scala combination.   Right now, this is only available thru `https://github.com/sparklyr/sparklyr/tree/main/inst/java`.  And, there is currently no external other ways of indicating the location for this jar other than via github. 

Is this about right? 

I noticed that https://github.com/sparklyr/sparklyr/tree/main/inst/java - does not have Spark 2.2 available. If Spark version livy configures with is spark 2.2 , then which version would it pick up instead?

**Comment by yitao-li** [comment]
@ying1 Yes that's correct.

**Comment by yitao-li** [comment]
@ying1 I was able to make the use case you described to work with the following steps:

- Run `install.packages("remotes")` if you haven't done so already
- `remotes::install_github("yitao-li/sparklyr", "fix_livy_issues")` to install a version of `sparklyr` that contains the fixes in https://github.com/sparklyr/sparklyr/pull/2779
- In the `apache-livy-0.7.0-incubating-bin/conf` directory (or <your livy installation>/conf) check if there is a text file named `livy.conf`. If not, then create an empty text file with that name.
- In `livy.conf`, insert or replace this line:
`livy.file.local-dir-whitelist = <comma separated list of paths that client is allowed to load jar files for their livy sessions>` -- I believe those paths are white-listed paths that exist on the Livy server
In my example (see below), I have `livy.file.local-dir-whitelist = /mnt/sda7/lib/R/library/sparklyr/java,/home/yitaoli/.ivy2/jars`
- Restart the Livy server (in my example I'm running Livy server locally, so, I just ran something like this 
 `SPARK_HOME=/home/yitaoli/spark/spark-2.4.4-bin-hadoop2.7 apache-livy-0.7.0-incubating-bin/bin/livy-server`
- Run `spark_connect` with multiple jar files specified in `conf$sparklyr.livy.jar`, e.g.,


```
library(sparklyr)

conf <- spark_config()
conf$sparklyr.livy.jar <- c(
  "/home/yitaoli/.ivy2/jars/org.sparklyr_sparklyr-flint_2-4_2-11-0.7.0.jar",
  "/mnt/sda7/lib/R/library/sparklyr/java/sparklyr-2.4-2.11.jar"
)

# This will work assuming the directories of both jar files above are white-listed
# in `ivy.file.local-dir-whitelist` in your livy.conf file
sc <- spark_connect(
  master = "http://localhost:8998", method = "livy", config = conf, version = "2.4"
)
```

Let me know if the above makes sense.

**Comment by ying1** [comment]
@yitao-li 
Thanks. Will give that a try in a bit.  
The file `/mnt/sda7/lib/R/library/sparklyr/java/sparklyr-2.4-2.11.jar"` - is it the same file as `https://github.com/sparklyr/sparklyr/blob/main/inst/java/sparklyr-2.4-2.11.jar` 

So in my setup - I'll just need to do 
```
conf$sparklyr.livy.jar <- c( . ...  required jars, "/mnt/sda7/lib/R/library/sparklyr/java/sparklyr-2.4-2.11.jar")
```

correct?

**Comment by yitao-li** [comment]
@ying1 Yes it's the same file as the one being downloaded from github.

I think currently the way it works is if `sparklyr` is installed in `/some_directory/lib/R/library/sparklyr` on your Livy server, then you will need to include `/some_directory/lib/R/library/sparklyr/java/sparklyr-2.4-2.11.jar`

Also re: your question about Spark version 2.2 -- your should try including the highest version of the sparklyr jar file available with version number less than or equal to your Spark version, so I think that would be the sparklyr-2.0 jar file in this case

**Comment by ying1** [comment]
@yitao-li 
Thank you.  With the new changes,  and also with the latest (43 minutes ago) https://github.com/sparklyr/sparklyr/blob/main/inst/java/sparklyr-2.4-2.11.jar  
I am able to pass the jars over to the remote system, and have the library loaded properly. 

Just as FYI.  Yesterday we picked up `sparklyr-2.4-2.11.jar` , and that was throwing NPE within Spark. This is no longer an issue in this build.

How often these jars change? or if it is tested per sparklyr release? (say 1.0.5).  Are these jar files distributed with sparklyr?   Let me know if I need to create a separate task for other questions, as I need to determine how on an enclosed system, we can reference these jars.

**Comment by yitao-li** [comment]
@ying1 The jar files form the main branch of `sparklyr` will have frequent changes, but `sparklyr` by default downloads those jar files from snapshot branches such as `feature/sparklyr-1.4.0` which contain jar files that are never changed after they are pushed.

I think the best thing to do here is to simply pick up the version of the jar file that works for your current use case and then store it in some artifact server within your infra. It's also advisable to store the GitHub URLs of those jar files that have commit hashes in case you need to re-download jar files from a particular commit from GitHub (e.g., https://github.com/sparklyr/sparklyr/blob/ea955cf76572d0386d0a288bd479e21cc045889d/inst/java/sparklyr-2.4-2.11.jar instead of .../blob/main/inst/...)

**Comment by yitao-li** [comment]
re: testing -- yes the jar files that are pushed to snapshot branches such as `feature/sparklyr-1.4.0` are indeed tested for each `sparklyr` release

**Comment by ying1** [comment]
@yitao-li Thank you.

**Comment by ying1** [comment]
@yitao-li 
What is the normal timeline for sparklyr releases?  I see that this fix is to be delivered for sparklyr 1.5.0

**Comment by yitao-li** [comment]
@ying1 Normally sparklyr is released to CRAN every 2-3 months . Sparklyr 1.5 release should happen very soon.
